pkg/container-runtime/configurator.go

```diff
@@ -0,0 +1,305 @@
+)
+
+//Configurator provide configuration Interface for different container runtime.
+type Configurator interface {
+	// ConfigRegistry config registry via each container runtime ip address
+	ConfigRegistry(RegistryConfig, []net.IP) error
+}
+
+type RegistryConfig struct {
+	SkipTLSVerify      bool
+	RegistryDeployHost string
+	Endpoint           string
+	Domain             string
+	Port               string
+	Username           string
+	Password           string
+	CaFile             string
+	KeyFile            string
+	CertFile           string
+}
+
+type DockerRuntimeConfigurator struct {
+	rootfs      string
+	infraDriver infradriver.InfraDriver
+}
+
+func (d DockerRuntimeConfigurator) ConfigRegistry(config RegistryConfig, ips []net.IP) error {
+	if err := d.configHostsFile(config.RegistryDeployHost, config.Domain, ips); err != nil {
+		return err
+	}
+
+	if err := d.configAuthInfo(config.Username, config.Password, config.Endpoint, ips); err != nil {
+		return err
+	}
+
+	if err := d.configRegistryCert(config.Domain, config.Port, config.CaFile, ips); err != nil {
+		return err
+	}
+
+	if err := d.configDaemon(config.Endpoint, ips); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d DockerRuntimeConfigurator) configHostsFile(registryIP, domain string, ips []net.IP) error {
+	// add registry ip to "/etc/hosts"
+	hostsLine := registryIP + " " + domain
+	writeToHostsCmd := fmt.Sprintf(RemoteAddEtcHosts, hostsLine, hostsLine)
+
+	f := func(host net.IP) error {
+		err := d.infraDriver.CmdAsync(host, writeToHostsCmd)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func (d DockerRuntimeConfigurator) configAuthInfo(username, password string, url string, ips []net.IP) error {
+	// docker login with username and password
+	// cp /root/.docker/config.json to /var/lib/kubelet. make sure kubelet could access registry with credential.
+	if username == "" || password == "" {
+		return nil
+	}
+
+	configAuthCmd := fmt.Sprintf(DockerLoginCommand, username, password, url)
+
+	f := func(host net.IP) error {
+		err := d.infraDriver.CmdAsync(host, configAuthCmd)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func (d DockerRuntimeConfigurator) configRegistryCert(domain, port string, caFile string, ips []net.IP) error {
+	// copy ca cert to "/etc/docker/certs.d/${domain}:${port}/${domain}.crt
+	src := filepath.Join(d.rootfs, "certs", caFile)
+	dest := filepath.Join(DockerCertDir, domain+":"+port, caFile)
+
+	if !os.IsFileExist(src) {
+		return nil
+	}
+
+	f := func(host net.IP) error {
+		err := d.infraDriver.Copy(host, src, dest)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return concurrencyExecute(f, ips)
+}
+
+func (d DockerRuntimeConfigurator) configDaemon(endpoint string, ips []net.IP) error {
+	// modify daemon.json: "mirrors": ["${endpoint}"],docker server need to restart if modify daemon.json
```

Comment:
No, you can just reload docker daemon configuration file rather than restart the daemon.